### PR TITLE
Add support to handle 402, 413, 414, 431 http error code as permanent errors in OTLP exporter #5674

### PR DESCRIPTION
**Pull request by mcmho**
**Description:** 
Collector currently does retries 402 and 413 error code. 413 is the error code for PAYLOAD TOO LARGE
and 402 is for PAYMENT REQUIRED which a user case of this would be when license token is exhausted.
These error code should be treated as permanent error in the Collector so that they won't get retried.

**Link to tracking Issue:** 
https://github.com/open-telemetry/opentelemetry-collector/issues/5674

**Testing:** 
oltp_test.go

**Documentation:** 
if we currently document http error code behavior, then treating 402 and 413 as permanent failure can be documented

### CODE REVIEW

**Comment by mcmho** [comment]
> @mcmho please add the appropriate unit test cases and CHANGELOG entry

okay. will do.

**Comment by mcmho** [comment]
unit test is there.    please help review 
cc @bogdandrutu  @pavankrish123

**Review by TylerHelmuth — APPROVED** [review]
LGTM

File: exporter/otlphttpexporter/otlp.go [hunk 1]
```diff
@@ -178,7 +178,9 @@ func (e *exporter) export(ctx context.Context, url string, request []byte) error
 		return exporterhelper.NewThrottleRetry(formattedErr, time.Duration(retryAfter)*time.Second)
 	}
 
-	if resp.StatusCode == http.StatusBadRequest {
+	// do not retry these errors
+	if resp.StatusCode == http.StatusBadRequest || resp.StatusCode == http.StatusRequestEntityTooLarge ||
```
[the remark this answers is not in the document]
Comment (mcmho) [reply exporter/otlphttpexporter/otlp.go:182]: @jpkrohling  Sure, I am down for clean code!   let me improve it!

**Review by jpkrohling — APPROVED** [review]

**Comment by jpkrohling** [comment]
You also need a changelog entry for this. I don't think this would qualify as a breaking change but as an enhancement.

File: exporter/otlphttpexporter/otlp.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pavankrish123) [reply exporter/otlphttpexporter/otlp.go]: @mcmho please see https://github.com/open-telemetry/opentelemetry-collector/issues/5697 and see if you can pick the 431 as well? We can use the same PR>

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,7 @@
 - Add `linux-ppc64le` architecture to cross build tests in CI
 - `client`: perform case insensitive lookups in case the requested metadata value isn't found (#5646)
 - `loggingexporter`: Decouple `loglevel` field from level of logged messages (#5678)
+- Add support to handle 402 and 413 http error code in Otlp exporter (#5685)
```
Comment (jpkrohling) [inline CHANGELOG.md:12]: Sorry for nitpicking, but OTLP needs to be in all caps:

```suggestion
- Add support to handle 402 and 413 http error code in OTLP exporter (#5685)
```

File: exporter/otlphttpexporter/otlp.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mcmho) [reply exporter/otlphttpexporter/otlp.go]: It's done.

**Comment by mcmho** [comment]
> Before approving, can you make sure this PR has all the cases that #5705 had?

Sure, it's done.

File: exporter/otlphttpexporter/otlp.go [hunk 3]
```diff
@@ -178,15 +178,44 @@ func (e *exporter) export(ctx context.Context, url string, request []byte) error
...
+	if code == http.StatusRequestURITooLong {
+		return true
+	}
+
+	// 431 - headers too large
+	if code == http.StatusRequestHeaderFieldsTooLarge {
+		return true
+	}
+
+	return false
```
Comment (bogdandrutu) [inline exporter/otlphttpexporter/otlp.go:216]: Use a switch statement?

Comment (mcmho) [reply exporter/otlphttpexporter/otlp.go:216]: The key original purpose of such structure is for code commenting.
Sure @bogdandrutu , I can make the change.     switch statement looks similar though:
```
	switch code {
            case http.StatusBadRequest:
		// 400 - bad request
		isPermanentError = true
            case http.StatusPaymentRequired:
		// 402 - payment required typically means that an auth token isn't valid anymore ...  
		isPermanentError = true
	    case http.StatusRequestEntityTooLarge:
		// 413 - request size is too large
		isPermanentError = true
	    case http.StatusRequestURITooLong:
		// 414 - uri is too long
		isPermanentError = true
	    case http.StatusRequestHeaderFieldsTooLarge:
		// 431 - headers too large
		isPermanentError = true
	    default:
		isPermanentError = false
	}

```

**Comment by mcmho** [comment]
@bogdandrutu  can you approve?

File: exporter/otlphttpexporter/otlp.go [hunk 4]
```diff
@@ -178,15 +178,41 @@ func (e *exporter) export(ctx context.Context, url string, request []byte) error
...
+		// 413 - request size is too large
+		isPermanentError = true
+	case http.StatusRequestURITooLong:
+		// 414 - uri is too long
+		isPermanentError = true
+	case http.StatusRequestHeaderFieldsTooLarge:
+		// 431 - headers too large
+		isPermanentError = true
+	default:
+		isPermanentError = false
```
[the remark this answers is not in the document]
Comment (mcmho) [reply exporter/otlphttpexporter/otlp.go:210]: @bogdandrutu  are you suggesting to remove the comment???   cc @jpkrohling

File: exporter/otlphttpexporter/otlp.go [hunk 5]
```diff
@@ -178,15 +178,41 @@ func (e *exporter) export(ctx context.Context, url string, request []byte) error
...
+		// Do not retry; report the failure as permanent if the server thinks the request is malformed.
 		return consumererror.NewPermanent(formattedErr)
 	}
 
 	// All other errors are retryable, so don't wrap them in consumererror.NewPermanent().
 	return formattedErr
 }
 
+func isPermanentClientFailure(code int) bool {
+	var isPermanentError bool
```
[the remark this answers is not in the document]
Comment (mcmho) [reply exporter/otlphttpexporter/otlp.go]: okay, sure!    done.

File: exporter/otlphttpexporter/otlp.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mcmho) [reply exporter/otlphttpexporter/otlp.go]: ok, improved comments.

**Review by jpkrohling — APPROVED** [review]

**Comment by mcmho** [comment]
Looks good now?  @bogdandrutu

**Review by bogdandrutu — APPROVED** [review]

**Comment by mcmho** [comment]
I think i have done everything now.    thanks @bogdandrutu  @jpkrohling  @TylerHelmuth for your approval!
Let's wait for the merge is ready.

**Comment by mcmho** [comment]
"First-time contributors need a maintainer to approve running workflows. 
6 expected and 1 successful checks"  

cc @jpkrohling  @bogdandrutu @TylerHelmuth

**Pull request merged**